feat(exporters): add user-agent enrichment (#4560)
open-telemetry/opentelemetry-specification: specification/protocol/exporter.md

The format of the header SHOULD follow [RFC 7231][rfc-7231]. The conventions used for specifying the OpenTelemetry SDK language and version are available in the [Resource semantic conventions][resource-semconv].

Exporters MAY expose a configuration option to append a product identifier to the User-Agent

> header as defined in [RFC 7231][rfc-7231]. Such option should not be available as

Reviewer: `SHOULD NOT` (even `MUST NOT`?)

Author: I agree with  `MUST NOT`. Thanks @carlosalberto :)

commit https://github.com/open-telemetry/opentelemetry-specification/pull/4560/commits/d7d96d72f7f792c37444293590d517c118e47b58

The format of the header SHOULD follow [RFC 7231][rfc-7231]. The conventions used for specifying the OpenTelemetry SDK language and version are available in the [Resource semantic conventions][resource-semconv].

Exporters MAY expose a configuration option to add a product identifier to the User-Agent header. The resulting User-Agent SHOULD include the exporter's default User-Agent string. The intention is to support an identifier for an OpenTelemetry SDK/Agent [distribution][opentelemetry-distribution]. Typically an exporter would *prepend* the given identifier to its own. For example:

```
MyDistribution/x.y.z OTel-OTLP-Exporter-Python/1.2.3

> ```

Author: done in https://github.com/open-telemetry/opentelemetry-specification/pull/4560/commits/abb5c1bbaac9eaa763ff150d0dfe0c7f80538b95

The format of the header SHOULD follow [RFC 7231][rfc-7231]. The conventions used for specifying the OpenTelemetry SDK language and version are available in the [Resource semantic conventions][resource-semconv].

> Exporters MAY expose a configuration option to add a product identifier to the User-Agent header. The resulting User-Agent SHOULD include the exporter's default User-Agent string. The intention is to support an identifier for an OpenTelemetry SDK/Agent [distribution][opentelemetry-distribution]. Typically an exporter would *prepend* the given identifier to its own. For example:

Reviewer 2: General question/comment, outside of the scope of this PR, but related!
The entire user-agent section seems loosely defined - there is no prescribed format, so some could do "OTel-OTLP-Exporter-Python/1.2.3" others could do "OpenTelemetry-OTLP-Exporter-Python/1.2.3" or something else. Would it be better if spec prescribe the exact format, including when Distribution adds its identifier? That'll make it easy for receivers to easily parse out information from this.

Author: the paragraph above says
````
OpenTelemetry protocol exporters SHOULD emit a User-Agent header to at a minimum identify the exporter, the language of its implementation, and the version of the exporter.
```

It sets a minimum of info to identify an exporter but it's not strict on the order of this info. I do agree if we have something a bit more specific would be useful. Specially for parsing it.

> Exporters MAY expose a configuration option to add a product identifier to the User-Agent header. The resulting User-Agent SHOULD include the exporter's default User-Agent string. The intention is to support an identifier for an OpenTelemetry SDK/Agent [distribution][opentelemetry-distribution]. Typically an exporter would *prepend* the given identifier to its own. For example:

Reviewer 3: Perhaps. I'm fine with that wording, but I *had* intentionally avoided being specific on *how* the user-agent string may be changed. My intent was to leave that to guidance from https://datatracker.ietf.org/doc/html/rfc7231#section-5.5.3

```
MyDistribution/x.y.z OTel-OTLP-Exporter-Python/1.2.3
```

> 
> 

Reviewer 3: ```suggestion

```

To fix my least-favourite markdown-lint error:

```
./specification/protocol/exporter.md:219 MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

The format of the header SHOULD follow [RFC 7231][rfc-7231]. The conventions used for specifying the OpenTelemetry SDK language and version are available in the [Resource semantic conventions][resource-semconv].

Exporters MAY expose a configuration option to append a product identifier to the User-Agent
header as defined in [RFC 7231][rfc-7231]. Such option MUST NOT be available as
an environment variable since is meant to be used by [distributions][opentelemetry-distribution]

> to append their product identifier along with the exporter's one.

Reviewer 4: Seems odd to me to explicitly call out no env vars but make this configurable in code. I assume the idea behind disallowing env support is to prevent users from setting this or overriding the distro, but if it's a config they can do that anyway. Should this be hidden entirely from users by e.g. making it a protected class property that a distro would override in a subclass?

Reviewer 2: even if Env variable exists, it cannot override the code based setting, as [code wins over env variable in almost all OTel implementations](https://github.com/open-telemetry/opentelemetry-specification/issues/4413).

Reviewer 5: Maybe instead of being precise about how people can interact with that (references to environment variables) we can just say that users (distro developers or even users) MAY change the user agent but the updated one MUST be superset of the OTel one?

Reviewer 3: How about the following wording. I think this satisfies the concerns from a number of comments here.

````suggestion
Exporters MAY expose a configuration option to add a product identifier to the User-Agent header. The resulting User-Agent SHOULD include the exporter's default User-Agent string. The intention is to support an identifier for an OpenTelemetry SDK/Agent [distribution][opentelemetry-distribution]. Typically an exporter would *prepend* the given identifier to its own. For example:

```
MyDistribution/x.y.z OTel-OTLP-Exporter-Python/1.2.3
```
````

> SDKs MAY expose a similar configuration option if exporters have it. This configuration

Author: SDK option removed in https://github.com/open-telemetry/opentelemetry-specification/pull/4560/commits/2451ae29c4a4fb8621f9b908cc2e4b8a5266a546

> SDKs MAY expose a similar configuration option if exporters have it. This configuration
> MUST be overridable by a signal specific option.
> 

Reviewer 3: Let's drop the mention of an *SDK* option and keep this limited to exporters.  If needed/wanted/helpful, it could be left to a *separate* PR/discussion on whether the *SDK* should expose this option.

```suggestion
```

As an example setting the option to `MyDistribution/x.y.z` in the exporter from the
above example would report the following:

```

> OTel-OTLP-Exporter-Python/1.2.3 MyDistribution/x.y.z

Reviewer 6: Should this be `MyDistribution/x.y.z OTel-OTLP-Exporter-Python/1.2.3` instead? The distribution (user's application) is typically more significant for identification than the underlying exporter library.

Also suggested here - https://www.rfc-editor.org/rfc/rfc9110#name-user-agents

```
User-Agent: CERN-LineMode/2.15 libwww/2.17b3
```

the first one is the application (browser), and second one is the library used by it.

> As an example setting the option to `MyDistribution/x.y.z` in the exporter from the
> above example would report the following:
> 
> ```
> OTel-OTLP-Exporter-Python/1.2.3 MyDistribution/x.y.z
> ```

Reviewer 3: I included an example (using *pre*pending rather than appending, as suggested) in a suggested-change comment above.

```suggestion
```

[otlp-http]: https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlphttp
[retryable-grpc-status-codes]: https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#failures
[retryable-http-status-codes]: https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#failures-1

> [opentelemetry-distribution]: https://github.com/open-telemetry/opentelemetry.io/edit/main/content/en/docs/concepts/distributions.md#what-is-a-distribution

Reviewer 3: ```suggestion
[opentelemetry-distribution]: https://opentelemetry.io/docs/concepts/distributions/
```
